Implements white-flag confirmation

luca-moser opened the pull request:
This PR implements the white-flag confirmation logic in accordance to RFC https://github.com/iotaledger/protocol-rfcs/blob/master/text/0005-white-flag/0005-white-flag.md

luca-moser committed:
implements white-flag

luca-moser committed:
adds white-flag merkle tree root hash to milestone head tx

luca-moser committed:
remove duped config option

muXxer commented:
:scam:

muXxer commented on pkg/whiteflag/white_flag.go:
```diff
@@ -0,0 +1,247 @@
+	}
+
+	// here we reached a tail of which its past cone was already visited, therefore we include its bundle
+	if (trunkVisited || tangle.SolidEntryPointsContain(headTxTrunkHash) || trunkConfirmedIndex != milestoneIndex) ||
```
Shouldn't trunk AND branch hit one of the 3 different criterias, for the bundle to be included?
What happens if branch is a solid entry point or confirmed by an older milestone, but not trunk?
I thought we first do a deep first search via trunk...

muXxer replied:
Another problem I see here, isn't trunkConfirmedIndex always 0 if not confirmed yet, and this way `trunkConfirmedIndex != milestoneIndex` will always be true? Same for branch.

muXxer replied:
Maybe we can move the traversal part:
```go

	if !tangle.SolidEntryPointsContain(headTxTrunkHash) && !trunkVisited && !trunkConfirmed {
		stack.PushFront(headTxTrunkHash)
		return nil
	}

	if !tangle.SolidEntryPointsContain(headTxBranchHash) && !branchVisited && !branchConfirmed {
		stack.PushFront(headTxBranchHash)
		return nil
	}
```
before that check?

luca-moser replied:
Yeah it was supposed to be and AND, and the confirmation check was substituted with the confirmed flag.

muXxer commented on pkg/whiteflag/white_flag.go:
```diff
@@ -0,0 +1,247 @@
+			// load state from milestone cone mutation or previous milestone
+			balance, has := wfConf.NewAddressState[addr]
+			if !has {
+				balanceStateFromPreviousMilestone, _, err := tangle.GetBalanceForAddressWithoutLocking(addr)
```
We should at least add a comment that we have to ReadLockLedger before calling `ProcessStack`

luca-moser committed:
adds exclusion sets

muXxer commented on pkg/whiteflag/white_flag.go:
```diff
@@ -135,29 +144,28 @@ func ProcessStack(stack *list.List, wfConf *Confirmation, visited map[trinary.Ha
 		defer cachedTrunkTx.Release()
-		trunkConfirmed, trunkConfirmedIndex = cachedTrunkTx.GetMetadata().GetConfirmed()
-	}
+		trunkConfirmed, _ = cachedTrunkTx.GetMetadata().GetConfirmed()
```
You could use the new function called `IsConfirmed`.

muXxer commented on pkg/whiteflag/white_flag.go:
```diff
@@ -135,29 +144,28 @@ func ProcessStack(stack *list.List, wfConf *Confirmation, visited map[trinary.Ha
 			}
 			defer cachedBranchTx.Release()
-			branchConfirmed, branchConfirmedIndex = cachedBranchTx.GetMetadata().GetConfirmed()
+			branchConfirmed, _ = cachedBranchTx.GetMetadata().GetConfirmed()
```
You could use the new function called `IsConfirmed`.

alexsporn committed:
- Converted whiteflag to binary hornet.Hash

alexsporn committed:
- Fixed comment

alexsporn commented on pkg/whiteflag/white_flag.go:
```diff
@@ -0,0 +1,265 @@
+	defer cachedBundle.Release()
+
+	if !cachedBundle.GetBundle().IsValid() || !cachedBundle.GetBundle().ValidStrictSemantics() {
+		return fmt.Errorf("%w: bundle %s is invalid", ErrMilestoneApprovedInvalidBundle, currentTx.Tx.Bundle)
```
@muXxer should we panic here? The RFC explicitely states

_In the event that an invalid bundle was encountered in the confirmation traversal, a node's expected behaviour would be to completely stop operations - and log the error - as it would mean the coordinator confirmed something it should not have._

alexsporn replied:
Or should we handle that in the solidifier and check for the ErrMilestoneApprovedInvalidBundle error?

alexsporn committed:
- Added test to the white flag merkle

- Added missing blake2b import or else it will panic at runtime about missing hash function

alexsporn committed:
- Suppress warnings since these values cannot be nil in these cases

alexsporn committed:
- Re-ordered imports

alexsporn committed:
- Also compute the address mutations when processing the whiteflag confirmation, not only the new ledger. We need this because we save the diff for each milestone to be able to reapply/revert each milestone on the ledger

alexsporn committed:
- Using whiteflag for confirmation

alexsporn committed:
- Removed the milestoneIndex from whiteflag since it is not used. In the pseudocode of the RFC it is used to check verify that the ledger mutation for the confirmation is done for the given milestone index, but in our case the transactions are not yet confirmed and are being confirmed by the current milestone, so the index makes no sense here.

alexsporn committed:
- Marking conflicting transactions and bundles in the metadata

- Confirming all tx referenced by a milestone, also zero value and conflicting

alexsporn committed:
- Do not trigger a ZMQ confirmation event for a conflicting tx

alexsporn committed:
- Do not trigger a MQTT confirmation event for a conflicting tx

alexsporn committed:
- Return false for conflicting tx in the getInclusionStates web api call

alexsporn committed:
- Added conflicting badge to the transaction explorer

alexsporn committed:
- Added missing byte to trytes conversion in some error messages

alexsporn commented:
Following tasks are still pending
- [ ] Mark conflicting tx in the visualizer
- [ ] Add conflicting, value and zerovalue tx to the Server Metrics graph

alexsporn committed:
Do not try to visit the bundle head branch and trunk tx if they are SEP

alexsporn committed:
- Added t6b1 package from RFC-0015 until this is integrated into iota.go

alexsporn committed:
- Made the whiteflag milestone merkle tree hash non-optional and made the hashing func used to compute it configurable. It defaults to BLAKE2b-512

alexsporn committed:
- Using the t6b1 encoding for writing the milestone merkle tree hash into the signature fragment of the milestone

alexsporn committed:
- Added GetMilestoneMerkleTreeHash() to bundle to extract the merkle tree hash from the milestone signature fragment

alexsporn committed:
- Verify the computed merkle tree hash from the milestone confirmation with the one contained in the milestone

alexsporn committed:
- Removed break from switch

alexsporn committed:
- Calm the doggo

alexsporn committed:
- Fixed test not having the BLAKE2b implementation

muXxer commented on pkg/metrics/serverMetrics.go:
```diff
@@ -18,6 +18,12 @@ type ServerMetrics struct {
 	KnownTransactions atomic.Uint32
 	// The number of confirmed transactions.
 	ConfirmedTransactions atomic.Uint32
+	// The number of conflicting transactions.
```
wrong comment

muXxer commented on pkg/metrics/serverMetrics.go:
```diff
@@ -18,6 +18,12 @@ type ServerMetrics struct {
 	ConfirmedTransactions atomic.Uint32
+	// The number of conflicting transactions.
+	ValueTransactions atomic.Uint32
+	// The number of conflicting transactions.
```
wrong comment

muXxer commented on pkg/model/tangle/bundle_milestone.go:
```diff
@@ -37,3 +39,20 @@ func (bundle *Bundle) GetMilestoneIndex() milestone.Index {
+	merkleRootHashSizeInTrytes := coordinatorMilestoneMerkleHashFunc.Size() * 6 / consts.TrinaryRadix
+	siblingsTrytesLength := coordinatorMerkleTreeDepth * consts.HashTrytesSize
+
+	if (siblingsTrytesLength + uint64(merkleRootHashSizeInTrytes)) > consts.KeyFragmentLength/consts.TrinaryRadix {
```
Why not use `SignatureMessageFragmentSizeInTrytes` instead?

muXxer commented on pkg/model/tangle/bundle_milestone.go:
```diff
@@ -37,3 +39,20 @@ func (bundle *Bundle) GetMilestoneIndex() milestone.Index {
+		return nil
+	}
+
+	merkleRootHashTrytes := headTx.GetTransaction().Tx.SignatureMessageFragment[:merkleRootHashSizeInTrytes]
```
Huh? I don't get that part, isn't the SignatureMessageFragment filled with `AuditPath` + `whiteFlagMerkleRootTreeHash`.

```diff
-merkleRootHashTrytes := headTx.GetTransaction().Tx.SignatureMessageFragment[:merkleRootHashSizeInTrytes]
+merkleRootHashTrytes := headTx.GetTransaction().Tx.SignatureMessageFragment[siblingsTrytesLength:siblingsTrytesLength+merkleRootHashSizeInTrytes]
```
Maybe also rename to `auditPathLength` or something?

muXxer commented on plugins/dashboard/frontend/src/app/components/ExplorerTransactionQueryResult.tsx:
```diff
@@ -196,7 +201,8 @@ export class ExplorerTransactionQueryResult extends React.Component<Props, any>
                                             </Link>
                                         </ListGroup.Item>
-                                        <ListGroup.Item className={style.monospace}>Obsolete Tag: {tx.obsolete_tag}</ListGroup.Item>
+                                        <ListGroup.Item className={style.monospace}>Obsolete
```
Is this change wanted?

muXxer commented on pkg/model/coordinator/coordinator.go:
```diff
@@ -67,19 +72,44 @@ type Coordinator struct {
+func MilestoneMerkleTreeHashFuncWithName(name string) crypto.Hash {
+	//TODO: golang 1.15 will include a String() method to get the string from the crypto.Hash, so we could iterate over them instead
+	var hashFunc crypto.Hash
+	switch name {
```
Maybe we should only check against lowercase name? So it would be less error prone to user mistakes.

muXxer commented on plugins/webapi/inclusion-states.go:
```diff
@@ -58,6 +58,10 @@ func getInclusionStates(i interface{}, c *gin.Context, _ <-chan struct{}) {
 		// check if tx is set as confirmed
 		confirmed := cachedTx.GetMetadata().IsConfirmed()
+		// Avoid passing true for conflicting tx to be backwards compatible
+		if confirmed && cachedTx.GetMetadata().IsConflicting() {
```
`confirmed := cachedTx.GetMetadata().IsConfirmed() && !cachedTx.GetMetadata().IsConflicting()`
Maybe in one statement.

muXxer commented on plugins/tangle/confirmation.go:
```diff
@@ -1,32 +1,55 @@
+	}
+
+	cachedMsTailTx := cachedMsBundle.GetBundle().GetTail()
+	defer cachedMsTailTx.Release()
```
You will release it twice, since all `cachedTxs` will be released anyways.

muXxer commented on plugins/tangle/confirmation.go:
```diff
@@ -1,32 +1,55 @@
 		// Release all bundles at the end
-		for _, cachedBndl := range cachedBndls {
-			cachedBndl.Release(true) // bundle -1
+		for _, cachedBundle := range cachedBundles {
```
lol :dagger:

alexsporn replied:
I will do a find&replace over all the codebase 😈

muXxer commented on plugins/tangle/confirmation.go:
```diff
@@ -35,136 +58,81 @@ func confirmMilestone(milestoneIndex milestone.Index, cachedMsTailTx *tangle.Cac
+			cachedBundle = tangle.GetCachedBundleOrNil(txHash) // bundle +1
+			if cachedBundle == nil {
+				cachedTx := loadTx(txHash)
+				log.Panicf("confirmMilestone: Tx: %v, Bundle not found: %v", txHash.Trytes(), cachedTx.GetTransaction().Tx.Bundle)
```
Maybe remove the loading of the cachedTx and just throw the bundle hash as error

muXxer commented on plugins/tangle/confirmation.go:
```diff
@@ -35,136 +58,81 @@ func confirmMilestone(milestoneIndex milestone.Index, cachedMsTailTx *tangle.Cac
-			}
-			cachedBndls[txHash] = cachedBndl
+	// we are only iterating over tail txs
+	onEachBundleTx := func(txHash hornet.Hash, do func(tx *tangle.CachedTransaction)) {
```
maybe rename to `forEachBundleTxWithTailTxHash`

muXxer commented on plugins/tangle/confirmation.go:
```diff
@@ -35,136 +58,81 @@ func confirmMilestone(milestoneIndex milestone.Index, cachedMsTailTx *tangle.Cac
-				log.Panicf("confirmMilestone: Tx: %v, Bundle not found: %v", hornet.Hash(txHash).Trytes(), cachedTx.GetTransaction().Tx.Bundle)
-			}
-			cachedBndls[txHash] = cachedBndl
+	// we are only iterating over tail txs
```
This completely confused me...
We loop over all txs of a specific bundle instance with tail tx hash

muXxer commented on plugins/tangle/confirmation.go:
```diff
@@ -35,136 +58,81 @@ func confirmMilestone(milestoneIndex milestone.Index, cachedMsTailTx *tangle.Cac
 	}
 
-	log.Infof("Milestone confirmed (%d): txsToConfirm: %v, collect: %v, total: %v", milestoneIndex, len(txsToConfirm), tc.Sub(ts), time.Since(ts))
+	log.Infof("Milestone confirmed (%d): txsConfirmed: %v, txsValue: %v, txsZeroValue: %v, txsConflicting: %v, collect: %v, total: %v", milestoneIndex, txsConfirmed, txsValue, txsZeroValue, txsConflicting, len(confirmation.TailsExcludedZeroValue), tc.Sub(ts), time.Since(ts))
```
Remove `len(confirmation.TailsExcludedZeroValue)`

muXxer commented on pkg/whiteflag/white_flag.go:
```diff
@@ -0,0 +1,282 @@
+			if cachedTrunkTx = tangle.GetCachedTransactionOrNil(headTxTrunkHash); cachedTrunkTx == nil {
+				return fmt.Errorf("%w: transaction %s", ErrMissingTransaction, headTxTrunkHash.Trytes())
+			}
+			defer cachedTrunkTx.Release()
```
maybe we should use `Release(true)`  here

muXxer commented on pkg/whiteflag/white_flag.go:
```diff
@@ -0,0 +1,282 @@
+	if cachedBundle == nil {
+		return fmt.Errorf("%w: bundle %s of candidate tx %s doesn't exist", ErrMissingBundle, currentTx.Tx.Bundle, currentTx.GetTxHash().Trytes())
+	}
+	defer cachedBundle.Release()
```
maybe we should use `Release(true)`  here

muXxer commented on pkg/whiteflag/white_flag.go:
```diff
@@ -0,0 +1,282 @@
+	if cachedTx == nil {
+		return fmt.Errorf("%w: candidate tx %s doesn't exist", ErrMissingTransaction, currentTxHash.Trytes())
+	}
+	defer cachedTx.Release()
```
maybe we should use `Release(true)`  here

muXxer commented on pkg/whiteflag/white_flag.go:
```diff
@@ -0,0 +1,282 @@
+	}
+
+	cachedBundleHeadTx := cachedBundle.GetBundle().GetHead()
+	defer cachedBundleHeadTx.Release()
```
maybe we should use `Release(true)`  here

muXxer commented on pkg/whiteflag/white_flag.go:
```diff
@@ -0,0 +1,282 @@
+				if cachedBranchTx = tangle.GetCachedTransactionOrNil(headTxBranchHash); cachedBranchTx == nil {
+					return fmt.Errorf("%w: transaction %s", ErrMissingTransaction, headTxBranchHash.Trytes())
+				}
+				defer cachedBranchTx.Release()
```
maybe we should use `Release(true)`  here

muXxer commented on pkg/whiteflag/white_flag.go:
```diff
@@ -0,0 +1,282 @@
+	}
+
+	if !tangle.SolidEntryPointsContain(headTxBranchHash) && !branchVisited && !branchConfirmed {
+		stack.PushFront(headTxBranchHash)
```
In case `bytes.Equal(headTxTrunkHash, headTxBranchHash)`, wouldn't we push the same transaction two times to the stack?

alexsporn committed:
- Requested changes from review

muXxer commented on pkg/model/tangle/bundle_milestone.go:
```diff
@@ -47,12 +47,12 @@ func (bundle *Bundle) GetMilestoneMerkleTreeHash() []byte {
+	auditPathLength := coordinatorMerkleTreeDepth * consts.HashTrytesSize
 
-	if (siblingsTrytesLength + uint64(merkleRootHashSizeInTrytes)) > consts.KeyFragmentLength/consts.TrinaryRadix {
+	if (auditPathLength + uint64(merkleRootHashSizeInTrytes)) > consts.SignatureMessageFragmentSizeInTrytes {
```
:rage:

alexsporn committed:
- Wen learn proper go.....

alexsporn commented:
Dashboard fixes will be tracked in #549
Tests will be tracked in #550
